Final/multinblkserver.cpp

```diff
@@ -0,0 +1,132 @@
+#include <unistd.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include <iostream>
+#include "ThreadPool.h"
+#include "LinkedList.h"
+
+const int PORT_NO = 8080;
+const int BACKLOG_SIZE = 2;
+
+// Prototypes
+void handle_connection(fd_set read_fds, LinkedList *open_conn_fds, ThreadPool *thread_pool);
+void echo(int fd);
+void accept_connection(int sockfd, LinkedList *open_conn_fds);
+
+void error(const char *msg)
+{
+    perror(msg);
+    exit(0);
+}
+
+int main(int argc, char const *argv[])
+{
+    // Step 1: Set up socket
+    int sockfd = socket(AF_INET, SOCK_STREAM, 0); // Create socket file descriptor for server
+    if (sockfd < 0)
+        error("ERROR opening socket");
+
+    // Step 2: Bind
+    struct sockaddr_in host;
+    host.sin_family = AF_INET; // IPv4 connection
+    host.sin_port = htons(PORT_NO); // bound to port 8080
+    host.sin_addr.s_addr = INADDR_ANY; // connected to localhost
+    const struct sockaddr *host_ptr = (struct sockaddr *) &host;
+    int bound = bind(sockfd, host_ptr, sizeof(host)); // bind sockfd to port
+    if (bound < 0)
+        error("ERROR binding to socket");
+
+    // Step 3: Listen
+    int listening = listen(sockfd, BACKLOG_SIZE); // The socket will actively listen for incoming connections and queue them up
+    if (listening < 0)
+        error("ERROR listener failed");
+
+    ThreadPool thread_pool(4); // create threadpool of size 4
+    LinkedList *open_conn_fds = new LinkedList();
+    while (1)
+    {
+        // Reset read_fds
+        fd_set read_fds;
+        FD_ZERO(&read_fds);
+        FD_SET(sockfd, &read_fds);
+        int maxfd = 0;
+        for (node *open_conn_fd = open_conn_fds->begin(); open_conn_fd; open_conn_fd = open_conn_fd->next)
```

Comment:
you shouldn't be relying on open_conn_fd != NULL or else you'll iterate through the sentinel node